README
::CalendarDateSelect::
Author:   Tim Harper ( 'tim_see_harperATgmail._see_om'.gsub('_see_', 'c').gsub('AT', '@') )

::Examples::
see demo here -- http://restatesman.com/static/calendar

::Project Site::
http://code.google.com/p/calendardateselect/

It works!



Setup for "vendored gem":
In enviroment.rb add the gem:
  config.gem "artmotion-calendar_date_select", :version => "1.10.6", :source => "http://gems.github.com", :lib => 'gem_init'

Install the gem locally (this installs the gem system wide)
	sudo rake gems:install
	
Unpack the gem locally (this unpacks it vendor/gems/ inside your applicatio)
	rake gems:unpack

Remove the system wide gem (this is require until Ticket#324 is closed [http://rails.lighthouseapp.com/projects/8994/tickets/324-gem-dependencies-inconsistently-load-rails-init-rb])
	sudo gem uninstall artmotion-calendar_date_select

Don't forget to include the javascript helper in your view:
	<%= javascript_include_tag :defaults %>  
	<%= calendar_date_select_includes %>

	

CHANGELOG
 * Calendar Date Select now works as a ruby-gem (thanks, artmotion!)

== Version 1.10.5 ==
Apr 7, 2008
 * Javascript code refactoring!  Less mess!
 * Bugfix - range was including element 42, which is really the 43rd element, so there was an error happening when today happens exactly 43 days after the beginning date (caught this by chance, surprised it wasn't reported before)

== Version 1.10.4 ==
Mar 24, 2008
 * patch applied to fix issue #92: nil object option causes CalendarDateSelect.calendar_date_select(object, method, options) to not use object.  Thanks, sskirby!

== Version 1.10.3 ==
Mar 24. 2008
 * Fixed active scaffold integration bug.  Thanks tapajos!

== Version 1.10.2 ==
Mar 10. 2008
 * bugfix: calendar_date_select_tag doesn't format Time and Date values (thanks for the patch, sskirby!)
 
== Version 1.10.1 ==
Mar 10, 2008
 * updated to automatically install locale folder
 * bugfix - cds "today" was showing up as Mar 9th 2008, when it was really Mar 10th 2008 - was caused by a wrinkle in time (Day light savings time)
 * Prototype 1.6.0.1 compatibility issue (77) resolved - thanks mike.nicholaides for the patch!
 * patch applied from mjroghelia to fix ie ssl iframe issue (84)
 * xhtml error fixed - &nbsp; replaced with #160;
 
==Version 1.10 ==
Dec 3, 2007
  * You can now pass in a parameter to tell CalendarDateSelect not to allow selection of certain dates.  See the [http://www.restatesman.com/static/calendar?tab=other demo] for an example.
  
==Version 1.9.2==
Nov 26, 2007
  * CalendarDateSelect was leaking javascript variables.  (nothing was explicitly declared local, so javascript assumes global).  Added test case and applied fix.  This may help with some of the conflicts CalendarDateSelect has been causing with other scripts.
  * Applied Ryan Wood's patch to fix the problem with using a form builder against objects like authorization[coverage_attributes][].  Thanks Ryan Wood!
  * Wes Hays brought up the point of parseInt trying to auto-detect number formats.  Wrote code to force interpretation of digits as base10, with test cases to prove it.
  * Misleading test case failure in Safari fixed: Safari won't let a boolean value to a property that should have a function (understandably).  Changed the test variable names to prevent such a conflict.
  * Now uses css :hover, rather than manually setting/removing the hover class.  If you are using a custom stylesheet, you'll need to change ".calendar_date_select tbody td.hover" to ".calendar_date_select tbody td:hover" (thanks, Alex Egg)
  
==Version 1.9.1==
Nov 15, 2007
  * Removed dependency on deprecated features in 1.6.0.  Thanks again, Richard Quadling!
  * Bugfix - was showing wrong day selected when selecting Mar 23, 2008.  (Day light savings time issue).  Fixed with test case.
  * Added CSS style to show red border around the embedded calendar_date_select div's with errors.
  * Callbacks were broken with 1.9.0.  Fixed, added test cases.

==Version 1.9.0==
Nov 13, 2007
  * Prototype 1.6.0 compatibility update.  Will now break with earlier versions of prototype (ie 1.5.1).  Big thanks to Kevin Alexander and Richard Quadling
  * Additional test coverage for improved stability going forward.

==Version 1.8.3==
Nov 12, 2007

	* Improved close methods:
	  * Added "OK" button at the bottom
		* Double clicking a day closes the calendar (not in IE, though... because IE is everyone's favorite browser to program for)
		* Escape key closes calendar
	* Added optional close button at the top, which is hidden by default.  Don't use this unless you really need it, because it may disappear in the future.
	* Don't focus a hidden element
	* Navigate down to child input element if exists (in case a div tag is passed in).  Resolves problems when input control is wrapped with a error div.  Test cases
	* Allow specific year ranges (pass in an array)
	  * restrict year navigation inside of a non-flexible range
	  * test case to check auto-repopulating of flexibleYearRanges.
	  * test case to check boundaries of nonFlexibleYearRanges.
		* Ability to pass in :year_range => 5.years.ago..0.years.ago to calendar_date_select (ruby)
	* Updated test cases
	* Bugfix for duplicate days that were occuring across day-light-saving-time boundaries
	* Added format_american.js (thanks, Wes Hays)
	* Added format_italian.js (thanks, Bigonazzi)
	* Added format_euro_24hr.js (thanks, James Silberbauer)
	* Bugfix for Konqueror - Was showing blank white box for popup window - had to do with Iframe hack.
	* Code-refactoring, clean up, breaking down methods, making the code easier to understand and extend.
	* Added new stylesheet definition: .calendar_date_select .cds_header a.close { display: none}.  Update your custom stylesheet with this new definition.
	
	
==Version 1.8.1==
September 27 2007
  * 1.8.0 was rushed.  There were a bunch of bugs that cropped up immediately.  This deploy includes JavaScript unit testing using the scriptaculous unit test framework! ([http://www.restatesman.com/calendar_date_select/js_test/functional/cds_test.html click here to run the tests on the latest trunk version]).  These unit tests are going to greatly improve the reliability of CalendarDateSelect releases.

==Version 1.8.0==

September 26 2007
  * Bugfix - Safari 2 issue fixed (was interpreting "" as January 1, 1969)
  * CSS overhaul - removed buttons and replaced with links.  Cleaned up CSS.  See CSSDiff_1_7_0__1_8_0 for a diff.
  * :popup => :force (force the user to select a date)
  * :time => "mixed" - You can now create a calendar control that allows a user to select "all day" or a specific time!
  * :month_year => "label" - Replace the clunky dropdowns with a text.
  * Some internal "gardening" on the code

==Version 1.7.1==

Aug 29 2007

  * Rendering options weren't being passed to the input control (like :style, :class, etc.)

==Version 1.7.0==

Aug 24 2007


public/javascripts/calendar_date_select.js
// CalendarDateSelect version 1.7.0 - a small prototype based date picker
// Questions, comments, bugs? - email the Author - Tim Harper <"timseeharper@gmail.seeom".gsub("see", "c")> 
if (typeof Prototype == 'undefined')
  alert("CalendarDateSelect Error: Prototype could not be found. Please make sure that your application's layout includes prototype.js (e.g. <%= javascript_include_tag :defaults %>) *before* it includes calendar_date_select.js (e.g. <%= calendar_date_select_includes %>).");

Element.addMethods({
  purgeChildren: function(element) { $A(element.childNodes).each(function(e){$(e).remove();}); },
  build: function(element, type, options, style) {
    newElement = Element.build(type, options, style);
    element.appendChild(newElement);
    return newElement;
  }
});

Element.build = function(type, options, style)
{
  e = $(document.createElement(type));
  $H(options).each(function(pair) { eval("e." + pair.key + " = pair.value" ); });
  if (style) 
    $H(style).each(function(pair) { eval("e.style." + pair.key + " = pair.value" ); });
  return e;
};
nil=null;

Date.one_day = 24*60*60*1000;
Date.weekdays = $w("S M T W T F S");
Date.months = $w("January February March April May June July August September October November December" );
Date.padded2 = function(hour) { padded2 = hour.toString(); if (parseInt(hour) < 10) padded2="0" + padded2; return padded2; }
Date.prototype.getPaddedMinutes = function() { return Date.padded2(this.getMinutes()); }
Date.prototype.getAMPMHour = function() { hour=this.getHours(); return (hour == 0) ? 12 : (hour > 12 ? hour - 12 : hour ) }
Date.prototype.getAMPM = function() { return (this.getHours() < 12) ? "AM" : "PM"; }
Date.prototype.stripTime = function() { return new Date(this.getFullYear(), this.getMonth(), this.getDate());};
Date.prototype.daysDistance = function(compare_date) { return Math.round((compare_date - this) / Date.one_day); };
Date.prototype.toFormattedString = function(include_time){
  str = Date.months[this.getMonth()] + " " + this.getDate() + ", " + this.getFullYear();
  
  if (include_time) { hour=this.getHours(); str += " " + this.getAMPMHour() + ":" + this.getPaddedMinutes() + " " + this.getAMPM() }
  return str;
}
Date.parseFormattedString = function(string) { return new Date(string);}
window.f_height = function() { return( [window.innerHeight ? window.innerHeight : null, document.documentElement ? document.documentElement.clientHeight : null, document.body ? document.body.clientHeight : null].select(function(x){return x>0}).first()||0); }
window.f_scrollTop = function() { return ([window.pageYOffset ? window.pageYOffset : null, document.documentElement ? document.documentElement.scrollTop : null, document.body ? document.body.scrollTop : null].select(function(x){return x>0}).first()||0 ); }

_translations = {
  "OK": "OK",
  "Now": "Now",
  "Today": "Today"
}
CalendarDateSelect = Class.create();
CalendarDateSelect.prototype = {
  initialize: function(target_element, options) {
    this.target_element = $(target_element); // make sure it's an element, not a string
    // initialize the date control
    this.options = $H({
      embedded: false,
      time: false,
      buttons: true,
      year_range: 10,
      calendar_div: nil,
      close_on_click: nil,
      minute_interval: 5,
      onchange: this.target_element.onchange
    }).merge(options || {});
    
    this.selection_made = $F(this.target_element)!="";
    
    if (this.target_element.calendar_date_select) { this.target_element.calendar_date_select.close(); return false; }
    this.target_element.calendar_date_select = this;
    this.callback("before_show")
    this.calendar_div = $(this.options['calendar_div']);
    if (!this.target_element) { alert("Target element " + target_element + " not found!"); return false;}
    
    this.parseDate();
    
    // by default, stick it by the target element (if embedded, that's where we'll want it to show up)
    if (this.calendar_div == nil) { this.calendar_div = $( this.options.embedded ? this.target_element.parentNode : document.body ).build('div'); }
    if (!this.options.embedded) {
      this.calendar_div.setStyle( { position:"absolute", visibility: "hidden" } )
      this.positionCalendarDiv();
    }
    
    this.calendar_div.addClassName("calendar_date_select");
    
    if (this.options["embedded"]) this.options["close_on_click"]=false;
    // logic for close on click
    if (this.options['close_on_click']===nil )
    {
      if (this.options['time'])
        this.options["close_on_click"] = false;
      else
        this.options['close_on_click'] = true;
    }
    
    // set the click handler to check if a user has clicked away from the document
    if(!this.options["embedded"]) Event.observe(document.body, "mousedown", this.bodyClick_handler=this.bodyClick.bindAsEventListener(this));
    
    this.initFrame();
    if(!this.options["embedded"]) { this.positionCalendarDiv(true) }//.bindAsEventListener(this), 1);
    this.callback("after_show")
  },
  positionCalendarDiv: function(post_painted) {
    above=false;
    c_pos = Position.cumulativeOffset(this.target_element); c_left = c_pos[0]; c_top = c_pos[1]; c_dim = this.calendar_div.getDimensions(); c_height = c_dim.height; c_width = c_dim.width; 
    w_top = window.f_scrollTop(); w_height = window.f_height();
    e_dim = Position.cumulativeOffset(this.target_element); e_top = e_dim[1]; e_left = e_dim[0];
    
    if ( (post_painted) && (( c_top + c_height ) > (w_top + w_height)) && ( c_top - c_height > w_top )) above=true;
    left_px = e_left.toString() + "px";
    top_px = (above ? (e_top - c_height ) : ( e_top + this.target_element.getDimensions().height )).toString() + "px";
    
    this.calendar_div.style.left = left_px;  this.calendar_div.style.top = top_px;
    
    // draw an iframe behind the calendar -- ugly hack to make IE 6 happy
    if (post_painted)
    {
      this.iframe = $(document.body).build("iframe", {}, { position:"absolute", left: left_px, top: top_px, height: c_height.toString()+"px", width: c_width.toString()+"px", border: "0px"})
      this.calendar_div.setStyle({visibility:""});
    }
  },
  initFrame: function() {
    that=this;
    // create the divs
    $w("header body time footer buttons").each(function(name) {
      eval(name + "_div = that." + name + "_div = that.calendar_div.build('div', { className: 'cds_"+name+"' }, { clear: 'left'} ); ");
    });
    
    this.initTimeDiv();
    this.initButtonsDiv();
    
    this.updateFooter("&nbsp;");
    // make the header buttons
    this.next_month_button = header_div.build("input", { type: "button", value : ">", className: "next" });
    this.prev_month_button = header_div.build("input", { type: "button", value : "<", className: "prev" });
    this.month_select = header_div.build("select", { className: "month"});
    this.year_select = header_div.build("select", { className: "year"});
    
    // make the month selector
    for(x=0; x<12; x++)
      this.month_select.options[x]=new Option(Date.months[x],x);
    Event.observe(this.prev_month_button, 'mousedown', function () { this.navMonth(this.date.getMonth() - 1 ) }.bindAsEventListener(this));
    Event.observe(this.next_month_button, 'mousedown', function () { this.navMonth(this.date.getMonth() + 1 ) }.bindAsEventListener(this));
    Event.observe(this.month_select, 'change', (function () { this.navMonth($F(this.month_select)) }).bindAsEventListener(this));
    Event.observe(this.year_select, 'change', (function () { this.navYear($F(this.year_select)) }).bindAsEventListener(this));
    
    // build the calendar day grid
    this.calendar_day_grid = [];
    days_table = body_div.build("table", { cellPadding: "0px", cellSpacing: "0px", width: "100%" }).build("tbody");

    // make the weekdays!
    weekdays_row = days_table.build("tr", {className: "weekdays"});
    Date.weekdays.each( function(weekday) { 
      weekdays_row.build("td", {innerHTML: weekday});
    });
    
    // Make the days!
    for(cell_index=0; cell_index<42; cell_index++)
    {
      weekday=cell_index%7;
      if ( cell_index %7==0 ) days_row = days_table.build("tr", {className: "days"});
      (this.calendar_day_grid[cell_index] = days_row.build("td", {
          calendar_date_select: this,
          onmouseover: function () { this.calendar_date_select.dayHover(this); },
          onmouseout: function () { this.calendar_date_select.dayHoverOut(this) },
          onclick: function() { this.calendar_date_select.updateSelectedDate(this); },
          className: (weekday==0) || (weekday==6) ? " weekend" : "" //clear the class
        },
        { cursor: "pointer" }
      )).build("div");
      this.calendar_day_grid[cell_index];
    }
    this.refresh();
  },
  initButtonsDiv: function()
  {
    buttons_div = this.buttons_div;
    if (!this.options["buttons"]) { Element.remove(buttons_div); return false; };
    
    b=buttons_div.build("input", {
      onclick: this.today.bindAsEventListener(this), 
      type: "button"
    });
    b.value = (this.options["time"] ? _translations["Now"] : _translations["Today"] );
    
    if (this.allowCloseButtons()) {
      b=buttons_div.build("input", {
        onclick: this.close.bindAsEventListener(this),
        type: "button"
      });
      b.value = _translations["OK"];
    }
  },
  initTimeDiv: function()
  {
    time_div = this.time_div;
    // make the time div
    if (this.options["time"])
    {
      time_div.build("span", {innerHTML:" @ "})
      this.hour_select = time_div.build("select", {
        calendar_date_select: this,
        onchange: function() { this.calendar_date_select.updateSelectedDate( { hour: this.value });}
      });
      time_div.build("span", {innerHTML:"&nbsp; : "});
      this.minute_select = time_div.build("select", {
        calendar_date_select: this,
        onchange: function() { this.calendar_date_select.updateSelectedDate( {minute: this.value }) }
      });
      t=new Date();
      // populate hours
      for(x=0;x<=23;x++) { t.setHours(x); this.hour_select.options[x] = new Option( "" + t.getAMPMHour() + " " + t.getAMPM(), x ) }
      // populate minutes
      x=0; for(m=0;m<60;m+=this.options["minute_interval"]) { this.minute_select.options[x++] = new Option( Date.padded2(m), m ) }
    } else (time_div.remove());
  },
  allowCloseButtons: function() { return ( !this.options["embedded"] && this.options["time"]); },
  dateString: function() {
    return (this.selection_made) ? this.selected_date.toFormattedString(this.options['time']) : "&nbsp;";
  },
  navMonth: function(month) {
    prev_day = this.date.getDate();
    this.date.setMonth(month);
    
    this.refresh();
    this.callback("after_navigate", this.date);
  },
  navYear: function(year) {
    this.date.setYear(year);
    this.refresh();
    this.callback("after_navigate", this.date);
  },
  refresh: function ()
  {
    // set the month
    this.month_select.selectedIndex = this.date.getMonth();
    
    // set the year, 
    range=this.options["year_range"];
    this.year_select.purgeChildren();
    for( x=0; x<=range*2; x++)
    {
      year = x+(this.date.getFullYear() - range);
      this.year_select.options[x]=new Option(year,year);
    }
    
    this.year_select.selectedIndex = range;
    
    // populate the calendar_day_grid
    this.beginning_date = new Date(this.date).stripTime();
    this.beginning_date.setDate(1);
    pre_days = this.beginning_date.getDay() // draw some days before the fact
    if (pre_days < 3) pre_days+=7;
    this.beginning_date.setDate(1 - pre_days);
    
    iterator = new Date(this.beginning_date);
    
    today = new Date().stripTime();
    this_month = this.date.getMonth();
    for (cell_index=0;cell_index<42; cell_index++)
    {

      day = iterator.getDate(); month = iterator.getMonth();
      cell = this.calendar_day_grid[cell_index];
      Element.remove(cell.childNodes[0]); div = cell.build("div", {innerHTML:day});
      if (month!=this_month) div.className = "other";
      cell.day=day; cell.month = month; cell.year = iterator.getFullYear();
      iterator.setDate( day + 1);
    }
    
    if (this.today_cell) this.today_cell.removeClassName("today");
    
    if ( $R(0,42).include(days_until = this.beginning_date.daysDistance(today)) ) {
      this.today_cell = this.calendar_day_grid[days_until];
      this.today_cell.addClassName("today");
    }
    
    // set the time
    if (this.options["time"]) {
      this.hour_select.selectedIndex = this.selected_date.getHours();
      this.minute_select.selectedIndex = this.selected_date.getMinutes() / this.options["minute_interval"];
      
      this.hour_select.onchange();
      this.minute_select.onchange();
    }
    
    this.setSelectedClass();
    this.updateFooter();
  },
  dayHover: function(element) {
    element.addClassName("hover");
    hover_date = new Date(this.selected_date);
    hover_date.setYear(element.year); hover_date.setMonth(element.month); hover_date.setDate(element.day);
    this.updateFooter(hover_date.toFormattedString(this.options['time']));
  },
  dayHoverOut: function(element) { element.removeClassName("hover"); this.updateFooter(); },
  setSelectedClass: function() {
    if (!this.selection_made) return;
    
    // clear selection
    if (this.selected_cell) this.selected_cell.removeClassName("selected");
    
    if ($R(0,42).include( days_until = this.beginning_date.daysDistance(this.selected_date.stripTime()) )) {
      this.selected_cell = this.calendar_day_grid[days_until];
      this.selected_cell.addClassName("selected");
    }
  },
  reparse: function() { this.parseDate(); this.refresh(); },
  parseDate: function()
  {
    this.date = Date.parseFormattedString(this.options['date'] || $F(this.target_element));
    if (isNaN(this.date)) this.date = new Date();
    this.selected_date = new Date(this.date);
    this.date.setDate(1);
  },
  updateFooter:function(text) { if (!text) text=this.dateString(); this.footer_div.purgeChildren(); this.footer_div.build("text", {innerHTML: text }); },
  updateSelectedDate:function(parts) {
    if (this.target_element.disabled || this.target_element.readOnly) return false;
    if (parts.day) {
      this.selection_made = true;
      for (x=0; x<=1; x++) {
      this.selected_date.setDate(parts.day);
      this.selected_date.setMonth(parts.month);
      this.selected_date.setYear(parts.year);}
    }
    
    if (parts.hour) this.selected_date.setHours(parts.hour);
    if (parts.minute) this.selected_date.setMinutes(parts.minute);
    
    this.updateFooter();
    this.setSelectedClass();
    
    if (this.selection_made) this.updateValue();
    if (this.options.close_on_click) { this.close(); }
    
  },
  updateValue: function() {
    last_value = this.target_element.value;
    this.target_element.value = this.dateString();
    if (last_value!=this.target_element.value) this.callback("onchange");
  },
  today: function() {
    this.date = new Date();
    d=new Date(); this.updateSelectedDate( { day: d.getDate(), month: d.getMonth(), year: d.getFullYear(), hour: d.getHours(), minute: d.getMinutes() } );
    this.refresh();
  },
  close: function() {
    this.callback("before_close");
    this.target_element.calendar_date_select = nil;
    Event.stopObserving(document.body, "mousedown", this.bodyClick_handler);
    this.calendar_div.remove();
    if (this.iframe) this.iframe.remove();
    this.target_element.focus();
    this.callback("after_close");
  },
  bodyClick: function(e) { // checks to see if somewhere other than calendar date select grid was clicked.  in which case, close
    if (! $(Event.element(e)).descendantOf(this.calendar_div) ) this.close();
  },
  callback: function(name, param) { if (this.options[name]) { this.options[name].bind(this.target_element)(param); } }
}